tensorflow_addons/losses/python/focal_loss.py

```diff
@@ -0,0 +1,142 @@
+"""Implements Focal loss"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import six
+import numpy as np
+
+import tensorflow as tf
+import tensorflow.keras.backend as K
+
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import nn
+from tensorflow.python.util.tf_export import keras_export
+from tensorflow.python.keras import losses
+from tensorflow_addons.utils.python import keras_utils
+
+
+@keras_export('keras.losses.SigmoidFocalCrossEntropy')
+class SigmoidFocalCrossEntropy(losses.LossFunctionWrapper):
+  """Focal loss down-weights well classified examples and focuses on the hard
+  examples. See https://arxiv.org/pdf/1708.02002.pdf for the loss definition.
+
+  Usage:
+
+  ```python
+  fl = tf.keras.losses.SigmoidFocalCrossEntropy()
+  loss = fl(
+    [[0.97], [0.91], [0.03]],
+    [[1], [1], [0])
+  print('Loss: ', loss.numpy())  # Loss: [[0.03045921]
+                                          [0.09431068]
+                                          [0.31471074]
+  ```
+  Usage with tf.keras API:
+
+  ```python
+  model = tf.keras.Model(inputs, outputs)
+  model.compile('sgd', loss=tf.keras.losses.SigmoidFocalCrossEntropy())
+  ```
+
+  Args
+    alpha: balancing factor, default value is 0.25
+    gamma: modulating factor, default value is 2.0
+  
+  Returns:
+    Weighted loss float `Tensor`. If `reduction` is `NONE`, this has the same
+        shape as `y_true`; otherwise, it is scalar.
+  
+  Raises:
+      ValueError: If the shape of `sample_weight` is invalid or value of 
+        `gamma` is less than zero
+
+
+  """
+
+  def __init__(self,
+               from_logits=False,
+               alpha=0.25,
+               gamma=2.0,
+               reduction=losses_utils.ReductionV2.NONE,
+               name='sigmoid_focal_crossentropy'):
+    super(SigmoidFocalCrossEntropy, self).__init__(
+        sigmoid_focal_crossentropy,
+        name=name,
+        reduction=reduction,
+        from_logits=from_logits,
+        alpha=alpha,
+        gamma=gamma)
+
+    self.from_logits = from_logits
+    self.alpha = alpha
+    self.gamma = gamma
+
+
+
+@keras_utils.register_keras_custom_object
+@tf.function
+def sigmoid_focal_crossentropy(y_true, 
+                                y_pred, 
+                                alpha=0.25,
+                                gamma=2.0,
+                                from_logits=False):
+    """
+    Args
+        y_true: true targets tensor.
+        y_pred: predictions tensor.
+        alpha: balancing factor.
+        gamma: modulating factor.
+    
+    Returns:
+        Weighted loss float `Tensor`. If `reduction` is `NONE`, this has the same
+        shape as `y_true`; otherwise, it is scalar.
+    """
+    if gamma:
+        if gamma < 0.: 
```

Comment:
mistake?

Reply:
Not a mistake but yeah the second if is redundant. Basically gamma can take a value of either `None` or anything greater than zero. I have combined the two ifs

Reply:
@facaiy My local branch is `add_new_loss`. Do you want me to merge my master with it?

Reply:
It's a good practice to pull from upstream and merge into your PR. See: http://blog.davidecoppola.com/2016/11/howto-contribute-to-open-source-project-on-github/ for setting up upstream, and the comments for people saying how to keep your branch up to date.

That said, I don't see any of your issues being because of a disconnect with upstream master.

Reply:
I have an upstream but it is merged with my master not the `add_new_loss` branch. I will merge the two